src/igraph_eulerian.c

```diff
@@ -0,0 +1,415 @@
+
+    igraph_stack_push(&tracker, start);
+
+    while (!igraph_stack_empty(&tracker)) {
+        
+        if (VECTOR(*outgoing_list)[curr] != 0) {
+            igraph_stack_push(&tracker, curr);
+            
+            igraph_inclist_init(graph, &il, IGRAPH_ALL);
+            incedges = igraph_inclist_get(&il, curr);
+            nc = igraph_vector_int_size(incedges);
+            edge = (long) VECTOR(*incedges)[0];
+            
+            next = IGRAPH_TO(graph, edge);
+
+            /* remove edge here */
+            VECTOR(*outgoing_list)[curr]--;
+            igraph_es_1(&es, edge);
+            igraph_delete_edges(graph, es);
+
+            curr = next;
+        } else { /* back track to find remaining circuit */
+            igraph_stack_push(&path, curr);
+            curr = igraph_stack_pop(&tracker);
+        }
+    }
+
+    while (!igraph_stack_empty(&path)) {
+        igraph_vector_push_back(res, igraph_stack_pop(&path));           
+    }
+
+    igraph_stack_destroy(&path);
+    igraph_stack_destroy(&tracker);
+
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+int igraph_eulerian_path_directed(igraph_t *graph, igraph_vector_t *res) {
+    igraph_es_t incoming, outgoing;
+    igraph_bool_t res_strong, res_weak;
+    igraph_integer_t incoming_excess, incoming_count, outgoing_excess, outgoing_count;
+    igraph_integer_t start_node;
+    igraph_vector_t outgoing_list;
+    igraph_t copy;
+    int i;
+
+    if (igraph_is_eulerian(graph) == 0) {
+        IGRAPH_WARNING("Euler cycle not possible");
+        return IGRAPH_FAILURE;
+    }
+
+    IGRAPH_CHECK(igraph_vector_init(&outgoing_list, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &outgoing_list);
+
+    IGRAPH_CHECK(igraph_copy(&copy, graph));
+    IGRAPH_FINALLY(igraph_destroy, &copy);
+
+    incoming_excess = 0;
+    outgoing_excess = 0;
+
+    /* determining the start node */
+    /* also getting the outgoing list for each vector*/
+    for (i = 0; i < igraph_vcount(graph); i++) {
+        igraph_es_incident(&incoming, i, IGRAPH_IN);
+        igraph_es_incident(&outgoing, i, IGRAPH_OUT);
+        
+        igraph_es_size(graph, &incoming, &incoming_count);
+        igraph_es_size(graph, &outgoing, &outgoing_count);
+
+        igraph_vector_push_back(&outgoing_list, outgoing_count);
+
+        if (incoming_count != outgoing_count) {
+            if ((incoming_count + 1 == outgoing_count) && (incoming_excess < 2 && outgoing_excess < 1)) {
+                outgoing_excess++;
+                start_node = i;
+                /* printf("start node is %d\n", start_node); */
+            } else if ((outgoing_count + 1 == incoming_count) && (incoming_excess < 1 && outgoing_excess < 2)) {
+                incoming_excess++;
+            } 
+        }
+    }
+
+    igraph_is_connected(graph, &res_strong, IGRAPH_STRONG);
+    igraph_is_connected(graph, &res_weak, IGRAPH_WEAK);
+
+    if ((outgoing_excess == 0 && incoming_excess == 0) && (res_strong == 1)) {
```

Comment:
If `b` is a boolean, please use `if (b)` and not `if (b == 1)`.

This is more readable, and generally "true" is represented with *a non-zero value* (not with 1).